### PR TITLE
Fork remote-cluster response handling

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today all responses to remote-cluster requests are deserialized and
handled on the transport worker thread. Some of these responses can be
sizeable, so with this commit we add the facility for callers to specify
a different executor to handle this work. It also adjusts several
callers to use more appropriate threads, including:

- responses from CCR-related admin actions are handled on `ccr`
- responses from field caps actions are handled on `search_coordination`

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java [hunk 1]
```diff
@@ -466,6 +466,7 @@ static void ccsRemoteReduce(
         ActionListener<SearchResponse> listener,
         BiConsumer<SearchRequest, ActionListener<SearchResponse>> localSearchConsumer
     ) {
+        final var remoteClientResponseExecutor = threadPool.executor(ThreadPool.Names.SEARCH_COORDINATION);
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:469]: I wonder if we should use the `SEARCH` pool instead, since that is the pool other work here is spawned on.

I think the coordination pool is mainly for the node level can-match/fieldcaps actions.

Might not matter a whole lot, but I'd rather induce any rejections due to this on the `SEARCH` thread pool than the `SEARCH_COORDINATOR` pool.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/Ccr.java [hunk 2]
```diff
@@ -373,10 +373,6 @@ public Optional<EngineFactory> getEngineFactory(final IndexSettings indexSetting
 
     @SuppressWarnings("HiddenField")
     public List<ExecutorBuilder<?>> getExecutorBuilders(Settings settings) {
-        if (enabled == false) {
```
Comment (henningandersen) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/Ccr.java:376]: Not that it matters, but was this change necessary or did you just find the "optimization" to not have the extra thread pool unnecessary?

**Review by henningandersen — COMMENTED** [review]
Looks good. I am in doubt about one of the thread pools though.

File: server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java:469]: I will revert this to SAME to make progress here, and raise it with the search folks to follow-up. I think what we're doing here is more like coordination work so enqueueing it behind some IO-heavy search activity might do bad things for performance, but there's definitely tradeoffs either way.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/Ccr.java (hunk 2 above)
Comment (DaveCTurner) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/Ccr.java:376]: We apparently do some CCR-related things even if CCR is disabled, or at least there are tests which now fail if the CCR threadpool is contingent on whether it's enabled or not. Seemed simplest to just create the threadpool either way rather than pick all that apart.

**Review by henningandersen — APPROVED** [review]
LGTM.

**Review by original-brownbear — APPROVED** [review]
LGTM2 nice :)

**Comment by DaveCTurner** [comment]
Thanks both :)
